cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -59,66 +57,69 @@ public SpendDeductZkp(IncentivePublicParameters incentivePublicParameters, Provi
+        builder.addSubprotocol(
+                "c0=usk*gamma+dsrnd0",
+                new LinearExponentStatementFragment(uskVar.mul(commonInput.gamma).add(dsrnd0Var).isEqualTo(commonInput.c0), zn)
+        );
 
         // c1=esk*gamma+dsrnd1
-        var c1statement = new ExponentSumExpr(new ExponentMulExpr(eskVar, new ExponentConstantExpr(commonInput.gamma)), dsrnd1Var);
-        builder.addSubprotocol("c1=esk*gamma+dsrnd1", new LinearExponentStatementFragment(c1statement.isEqualTo(commonInput.c1), zn));
+        builder.addSubprotocol(
+                "c1=esk*gamma+dsrnd1",
+                new LinearExponentStatementFragment(eskVar.mul(commonInput.gamma).add(dsrnd1Var).isEqualTo(commonInput.c1), zn)
+        );
 
         // dsid=w^esk
         var dsidEskStatement = w.pow(eskVar).isEqualTo(commonInput.dsid);
         builder.addSubprotocol("dsid=w^esk", new LinearStatementFragment(dsidEskStatement));
 
         // C=(H.pow(usk, esk, dsrnd_0, dsrnd_1, v, z, t),g_1) split into two subprotocols
-        var commitmentC0Statement = H.pow(new Vector<>(
+        var commitmentC0Statement = H.innerProduct(new Vector<>(
                 uskVar,
                 eskVar,
                 dsrnd0Var,
                 dsrnd1Var,
                 vVar,
                 zVar,
                 tVar
-        )).reduce(GroupElementExpression::op).isEqualTo(commonInput.commitmentC0);
+        )).isEqualTo(commonInput.commitmentC0);
         builder.addSubprotocol("C0", new LinearStatementFragment(commitmentC0Statement));
-        builder.addSubprotocol("C1", new LinearStatementFragment(commonInput.commitmentC1.expr().isEqualTo(pp.getG1())));
+        // C1=g is not sent and verified since no witness is involved.
 
         // C=(H.pow(usk, \sum_i=0^k[esk^*_(usr,i) * base^i], dsrnd^*_0, dsrnd^*_1, v-k, z^*, t^*), g_1^(u^*)) split into two subprotocols
         // We use the sum to combine the esk^*_usr = \sum proof with the C=.. proof
-        var exponents = new Vector<>(uskVar, BigInteger.ZERO, dsrndStar0Var, dsrndStar1Var, vVar.sub(zn.valueOf(commonInput.k)), zStarVar, tStarVar); // 0 because h_2^(esk^*_usr) is handled separately
-        GroupElementExpression eskPart = groupG1.getNeutralElement().expr();
-        for (int i = 0; i < pp.getNumEskDigits(); i++) {
-            eskPart = eskPart.op(H.get(1).pow(eskDecVarVector[i].mul(pp.getEskDecBase().pow(BigInteger.valueOf(i)))));
-        }
-
-        var cPre0Statement = H.pow(exponents).reduce(GroupElementExpression::op).op(eskPart).isEqualTo(commonInput.c0Pre.pow(uStarInverseVar));
+        var powersOfEskDecBase = new ExponentExpressionVector(
+                IntStream.range(0, pp.getNumEskDigits())
+                        .mapToObj(i -> pp.getEskDecBase().pow(BigInteger.valueOf(i)).asExponentExpression())
+                        .collect(Collectors.toList())
+        ); // construct vector (eskBase^0, eskBase^1, ...)
```

Comment:
This can also be simplified using `ExponentExpressionVector.generate` if you like.